<!DOCTYPE html>
<html lang="en">
  
  <head>
    <meta charset="utf-8" />
    <title>Bibliography notes</title>
    <link rel="stylesheet" href="./theme/css/main.css" />
    <!--[if IE]>
        <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->
  </head>

  <body id="index" class="home">
    <header id="banner" class="body">
      <h1><a href="./">Roscoff course 2015 - RAD tags </a></h1>
      <nav><ul>
<!-- http://stackoverflow.com/questions/18520046/how-can-i-control-the-order-of-pages-from-within-a-pelican-article-category -->
          <li><a href="./index.html">About</a></li>
          <li><a href="./schedule.html">Course schedule</a></li>
          <li class="active"><a href="./bibliography-notes.html">Bibliography notes</a></li>
          <li><a href="./draft-notes.html">Draft notes</a></li>
      </ul></nav>
    </header><!-- /#banner -->

<section id="content" class="body">
    <!-- <h1 class="entry-title">Bibliography notes</h1> -->
    
    


<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#sec-1">Miller (2007) - Genome Research (<b>*</b>)</a></li>
<li><a href="#sec-2">Miller (2007) - Genome Biology (*)</a></li>
<li><a href="#sec-3">Lewis (2007) - Genetics</a></li>
<li><a href="#sec-4">Baird (2008) - PLoS One (**)</a></li>
<li><a href="#sec-5">Hohenlohe (2010) - PLoS Genetics (<b>*</b>)</a></li>
<li><a href="#sec-6">Hohenlohe (2011) - Molecular Ecology Resources (*)</a></li>
<li><a href="#sec-7">Emerson (2011) - PNAS (**)</a></li>
<li><a href="#sec-8">Etter (2011) - PLoS One</a></li>
<li><a href="#sec-9">Etter (2011) - Methods in Molecular Biology (Rev)</a></li>
<li><a href="#sec-10">Davey (2011) - Briefings in Functional Genomics (Rev)</a></li>
<li><a href="#sec-11">Davey (2011) - Nature Reviews Genetics (Rev) (*)</a></li>
<li><a href="#sec-12">O'Rourke (2011) - Genetics</a></li>
<li><a href="#sec-13">Pfender (2011) - Theor. Appl. Genet. (*)</a></li>
<li><a href="#sec-14">Chong (2011) - Bioinformatics</a></li>
<li><a href="#sec-15">Flight (2012) - Integrative and Comparative Biology (*)</a></li>
<li><a href="#sec-16">Hohenlohe (2012) - Phil. Trans. R. Soc. B (**)</a></li>
<li><a href="#sec-17">Houston (2012) - BMC Genomics (*)</a></li>
<li><a href="#sec-18">Peterson (2012) - PLoS One (<b>*</b>)</a></li>
<li><a href="#sec-19">Rubin (2012) - PLoS One (**)</a></li>
</ul>
</div>
</div>
<p>
The idea of using restriction site associated DNA (RAD) markers on a large
scale was first introduced in a paper by <a href="http://genome.cshlp.org/content/17/2/240.long">Miller (2007)</a> from Eric
A. Johnson's group.
</p>
<p>
There are important differences between methods based on <b>restriction site
polymorphism</b> (like <a href="http://genome.cshlp.org/content/17/2/240.long">Miller 2007</a>) and methods that use restriction sites as a
way to <b>decrease genomic complexity</b> before sequencing (RAD-seq). While
restriction site polymorphism was historically (in 2007!) used first, mainly
the RAD-seq method and its derivative are used at present.
</p>
<p>
Below are bibliography notes related to the use of RAD tags - (\*), (\*\*) and
(\*\*\*) denotes papers of particular interest, and (Rev) denotes reviews.
</p>
<div class="outline-2" id="outline-container-sec-1">
<h2 id="sec-1">Miller (2007) - Genome Research (<b>*</b>)</h2>
<div class="outline-text-2" id="text-1">
<p>
<b><a href="http://genome.cshlp.org/content/17/2/240.long">Rapid and cost-effective polymorphism identification and genotyping using
restriction site associated DNA (RAD) markers</a></b>
</p>
<ul class="org-ul">
<li>Summary
<ul class="org-ul">
<li>Eric A. Johnson's group at the University of Oregon
</li>
<li>Definition of RAD tags: "Restriction site associated DNA (RAD) tags are a
<b>genome-wide representation of every site of a particular restriction enzyme
by short DNA tags</b>"
</li>
<li>Here, looking for polymorphism that disrupt restriction sites (like RFLP,
AFLP) but for many more loci than the previous methods.
</li>
<li>Not sequencing but use of a micro-array platform
</li>
<li>"<b>RAD markers […] allow high-throughput, high-resolution genotyping in
both model and nonmodel systems</b>"
</li>
</ul>
</li>
<li>Interest of RAD tags
<ul class="org-ul">
<li>Genetic mapping (recombination mapping, association mapping) needs
"high-throughput, cost-effective methods to identify and genotype
polymorphisms"
</li>
<li>Previous methods:
<ul class="org-ul">
<li>allele-specific oligonucleotide microarrays (thousands of SNPs scored per
hybridization)
</li>
<li>direct hybridization of labelled genomic DNA to short oligonucleotide
arrays
</li>
<li>single feature polymorphisms (SFP) can be discovered and typed in a similar
way
</li>
</ul>
</li>
<li>Those methods need costly preliminary steps (sequencing, oligonucleotide
synthesis) and are not easily feasible for non-model organisms
</li>
</ul>
</li>
<li>Principle
<ul class="org-ul">
<li>Very good illustration in <a href="http://genome.cshlp.org/content/17/2/240.long#F1">Figure 1</a>.
</li>
<li>Digest - link tags - shear DNA - purify tags - release tags - type tags
</li>
</ul>
</li>
<li>Demonstration
<ul class="org-ul">
<li>Mapping of a breakpoint on a <i>Drosophila</i> chromosome
</li>
<li>Identification of the major lateral plate locus in three-spine stickleback
</li>
</ul>
</li>
<li>Discussion
<ul class="org-ul">
<li>Here RAD markers use the <b>restriction site as the polymorphic marker</b> (no
sequencing)
</li>
<li>Easy to produce microarray type of genotyping resource
</li>
<li>Bulk genotyping is possible
</li>
<li>If markers location is determined, genetic mapping is possible
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-2">
<h2 id="sec-2">Miller (2007) - Genome Biology (*)</h2>
<div class="outline-text-2" id="text-2">
<p>
<b><a href="http://genomebiology.com/content/8/6/R105">RAD marker microarrays enable rapid mapping of zebrafish mutations</a></b>
</p>
<ul class="org-ul">
<li>Developed a RAD marker microarray resource for the zebrafish
</li>
<li>Bulk segregant approach to map mutations
</li>
<li>Assay for individual RAD markers in pooled populations
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-3">
<h2 id="sec-3">Lewis (2007) - Genetics</h2>
<div class="outline-text-2" id="text-3">
<p>
<b><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2034621/">High-Density Detection of Restriction-Site-Associated DNA Markers for Rapid
Mapping of Mutated Loci in Neurospora</a></b>
</p>
<ul class="org-ul">
<li>Micro-array based RAD markers in <i>Neurospora crassa</i>
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-4">
<h2 id="sec-4">Baird (2008) - PLoS One (**)</h2>
<div class="outline-text-2" id="text-4">
<p>
<b><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0003376">Rapid SNP Discovery and Genetic Mapping Using Sequenced RAD Markers</a></b>
</p>
<ul class="org-ul">
<li>RAD tags associated with Illumina sequencing for SNP discovery
</li>
<li>SNP identified by sequenced polymorphism and disrupted restriction sites
</li>
<li>Marker density can be tuned by the choice of the enzymes (and their number)
</li>
<li>Samples can be multiplexed by barcoding
</li>
<li>Useful with or without a reference genome
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-5">
<h2 id="sec-5">Hohenlohe (2010) - PLoS Genetics (<b>*</b>)</h2>
<div class="outline-text-2" id="text-5">
<p>
<b><a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000862">Population Genomics of Parallel Adaptation in Threespine Stickleback using
Sequenced RAD Tags</a></b>
</p>
<p>
A major paper about sequenced RAD tags.
</p>
<ul class="org-ul">
<li>Use natural populations, not lab crosses
</li>
<li>Genome scan of nucleotide diversity and differentiation in 5 natural
populations of threespine stickleback
</li>
<li>Illumina-sequenced RAD tags
</li>
<li>Identify and type more than 45000 SNPs in each of 100 individuals
</li>
<li>Genomic regions with balancing and divergent selection, consistent across
populations
</li>
<li>Identify the genes in regions of interest (regions of parallel
differentiation across independent populations)
</li>
<li>Methods:
<ul class="org-ul">
<li>Genome is available
</li>
<li>Barcoded sample
</li>
<li>Kernel smoothing and permutation testing
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-6">
<h2 id="sec-6">Hohenlohe (2011) - Molecular Ecology Resources (*)</h2>
<div class="outline-text-2" id="text-6">
<p>
<b><a href="http://onlinelibrary.wiley.com/doi/10.1111/j.1755-0998.2010.02967.x/abstract">Next-generation RAD sequencing identifies thousands of SNPs for assessing
 hybridization between rainbow and westslope cutthroat trout</a></b>
</p>
<ul class="org-ul">
<li>Nonmodel organism (no genome available)
</li>
<li>Salmonid fish, several whole-genome duplication
</li>
<li>RAD sequencing to identify SNP loci with fixed allelic differences between
introduced rainbow trout and native westslope cutthroat trout
</li>
<li>Use 24 barcode-labelled individuals
</li>
<li>Use Stacks for analysis (genotyping)
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-7">
<h2 id="sec-7">Emerson (2011) - PNAS (**)</h2>
<div class="outline-text-2" id="text-7">
<p>
<b><i>Resolving postglacial phylogeography using high-throughput
sequencing</i></b>
</p>
<ul class="org-ul">
<li>Fine-scale phylogenetic relationships between recently diverged (from a
refugium after the recession of the Laurentide ice sheet 22000-19000 BP)
populations of pitcher plant mosquito
</li>
<li>Discussed also in <i>Puritz 2012</i> ("Next-Generation Phylogeography: A Targeted
Approach for Multilocus Sequencing of Non-Model Organisms") in which they use
454 sequencing:
<blockquote>
<p>
"However, few of [the NGS] technologies have been applied in phylogeographic
studies, with the exception of Emerson et al. (42) that used
restriction-site-associated DNA tags (RAD tags) (43,44) to determine the
evolutionary relationship between recently diverged populations of pitcher
plant mosquitos. This methodology can genotype multiple populations at
thousands of SNP loci simultaneously, but have limited ability to survey
large sample sizes within populations because of the cost. For example,
Emerson et al. genotyped 21 different population at ,13,627 different SNPs
but with only 6 individuals per population the authors were forced to
restrict their analysis to 3,741 SNPs ‘‘that were fixed within at least two
populations and were variable among populations’’ and generate one consensus
sequence per locus per population (42). Our targeted 454 sequencing
methodology offers a compromise with the ability to sequence a reasonable
sample size (20 individuals) from one population for 10 different nDNA loci
(,400 bp each) in 1/16th of a plate of 454 sequencing, at lower cost than a
RAD tag analysis."
</p>
</blockquote>
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-8">
<h2 id="sec-8">Etter (2011) - PLoS One</h2>
<div class="outline-text-2" id="text-8">
<p>
<b><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0018561">Local De Novo Assembly of RAD Paired-End Contigs Using Short Sequencing Reads</a></b>
</p>
<ul class="org-ul">
<li>Use paired-end sequencing of RAD fragments to build contigs on the side which
is cut by random shearing
</li>
<li>Identify SNP and determine haplotype in threespine stickleback
</li>
<li>Produce overlapping contigs of several hundred nucleotides in <i>E. coli</i> and
threespine stickleback
</li>
<li>A circularization step can allow a local assembly of up to 5 kb
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-9">
<h2 id="sec-9">Etter (2011) - Methods in Molecular Biology (Rev)</h2>
<div class="outline-text-2" id="text-9">
<p>
<b><i>SNP Discovery and Genotyping for Evolutionary Genetics Using RAD Sequencing</i></b>
</p>
<ul class="org-ul">
<li>Review chapter about RAD genotyping (RAD with Illumina sequencing) or
  <b>RAD-seq</b>
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-10">
<h2 id="sec-10">Davey (2011) - Briefings in Functional Genomics (Rev)</h2>
<div class="outline-text-2" id="text-10">
<p>
<b><i>RADSeq: next-generation population genetics</i></b>
</p>
<ul class="org-ul">
<li>Review about RAD-seq and its use for population genomics
</li>
<li>Interest for non-model organisms, wild populations and ecological population
genomics
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-11">
<h2 id="sec-11">Davey (2011) - Nature Reviews Genetics (Rev) (*)</h2>
<div class="outline-text-2" id="text-11">
<p>
<b><i>Genome-wide genetic marker discovery and genotyping using
next-generation sequencing</i></b>
</p>
<ul class="org-ul">
<li>Review about NGS methods that use restriction enzyme digestion to reduce the
genome complexity
</li>
<li>Include reduced-representation sequencing using reduced-representation
libraries (RRLs) or complexity reduction of polymorphic sequences (CRoPS)
</li>
<li>Include RAD-seq
</li>
<li>Include low coverage genotyping
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-12">
<h2 id="sec-12">O'Rourke (2011) - Genetics</h2>
<div class="outline-text-2" id="text-12">
<p>
*<i>Rapid Mapping and Identification of Mutations in <i>Caenorhabditis elegans</i> by
Restriction Site-Associated DNA Mapping and Genomic Interval Pull-Down
Sequencing</i>*
</p>
<ul class="org-ul">
<li>Forward genetic screens in <i>C. elegans</i> (association between phenotypes and
mutated genes)
</li>
<li>Use RAD-seq for rapid mapping of the mutations
</li>
<li>Use genomic interval pull-down sequencing (GIPS) to capture and sequence
large portions of a mutant genome (ca. Mb)
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-13">
<h2 id="sec-13">Pfender (2011) - Theor. Appl. Genet. (*)</h2>
<div class="outline-text-2" id="text-13">
<p>
<b><i>Mapping with RAD (restriction-site associated DNA) markers to rapidly identify
QTL for stem rust resistance in <i>Lolium perenne</i></i></b>
</p>
<ul class="org-ul">
<li>Use RAD-seq to produce maps and identify QTL (in association with simple
sequence repeats (SSR) and sequence-tagged sites (STS))
</li>
<li>Reads from parental plants (2) and from F1 (188)
</li>
<li>High variability in the number of reads per F1 (14000 to 935000)
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-14">
<h2 id="sec-14">Chong (2011) - Bioinformatics</h2>
<div class="outline-text-2" id="text-14">
<p>
<b><i>Rainbow: an integrated tool for efficient clustering and
assembling RAD-seq reads</i></b>
</p>
<ul class="org-ul">
<li>Tools to assemble RAD reads
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-15">
<h2 id="sec-15">Flight (2012) - Integrative and Comparative Biology (*)</h2>
<div class="outline-text-2" id="text-15">
<p>
<b><i>Genetic Variation in the Acorn Barnacle from Allozymes
to Population Genomics</i></b>
</p>
<ul class="org-ul">
<li>Use allozymes, microsatellites, mtDNA and RADseq
</li>
<li>RADseq on pooled samples (20 individuals per pool) of 3 populations (2 US, 1
UK)
</li>
<li>Use SOAPdenovo assembly
</li>
<li>Fst outlier analysis (using unbiased estimates and empirical distribution,
with the 1% tail of the distribution being the outliers)
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-16">
<h2 id="sec-16">Hohenlohe (2012) - Phil. Trans. R. Soc. B (**)</h2>
<div class="outline-text-2" id="text-16">
<p>
<b><i>Extensive linkage disequilibrium and parallel adaptive divergence across
threespine stickleback genomes</i></b>
</p>
<ul class="org-ul">
<li>Study patterns of local and long-distance linkage desequilibrium (LD) across
oceanic and freshwater populations of threespine stickleback
</li>
<li>Look for association between LD and signatures of divergent selection
</li>
<li>Assess the role of recombination rate variation in generating LD patterns
</li>
<li>Reuse RAD-seq data from Hohenlohe 2010 (PLoS Genetics) and also generate new
data with a modified RAD-seq protocol on 87 F2 individuals and the parents
and F1 individuals
</li>
<li>Align raw reads to reference genome with Bowtie
</li>
<li>Use likelihood methods to infer the genotype at each nucleotide
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-17">
<h2 id="sec-17">Houston (2012) - BMC Genomics (*)</h2>
<div class="outline-text-2" id="text-17">
<p>
<b><i>Characterisation of QTL-linked and genome-wide restriction site-associated DNA
(RAD) markers in farmed Atlantic salmon</i></b>
</p>
<ul class="org-ul">
<li>RAD-seq on two Atlantic salmon families
</li>
<li>Offspring were classified into resistant and susceptible to a disease, and
for each family DNA from the parents and 7 offsprings of each phenotype was
sequenced by RAD-seq in multiplexed pools
</li>
<li>Use of pedigreed samples allowed to distinguish segregating SNPs from
putative paralogous sequences (recent genome duplication in salmonids)
</li>
<li>50 SNPs linked to QTL and a subset used for high-throughput genotyping across
large commercial populations of disease-challenged fry
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-18">
<h2 id="sec-18">Peterson (2012) - PLoS One (<b>*</b>)</h2>
<div class="outline-text-2" id="text-18">
<p>
<b><i>Double Digest RADseq: An Inexpensive Method for <i>De Novo</i> SNP Discovery and
Genotyping in Model and Non-Model Species</i></b>
</p>
<ul class="org-ul">
<li>Introduce the double-digest method
</li>
<li>Combinatorial barcoding for larger number of individuals per multiplex
</li>
<li>Detailed pipeline of reads analysis
</li>
<li>Also compare the observed size distribution with <i>in silico</i> digestion and
statistical models of size selection
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-19">
<h2 id="sec-19">Rubin (2012) - PLoS One (**)</h2>
<div class="outline-text-2" id="text-19">
<p>
<b><i>Inferring Phylogenies from RAD Sequence Data</i></b>
</p>
<ul class="org-ul">
<li>Simulate RAD-seq data collection for interspecific phylogeny reconstruction
in <i>Drosophila</i>, mammals and yeast
</li>
<li>Develop a workflow to test if informative data can be gained from RAD-seq and
used for accurate reconstruction of "known" phylogenies
</li>
<li>RAD-seq seems promising for phylogenies in younger clades (e.g. 60 My vs 100
My) in which enough orthologous restriction sites are retained across species
</li>
</ul>
</div>
</div>

</section>

    <!-- <section id="extras" class="body"> -->
    <!--  -->
    <!--  -->
    <!-- </section><\!-- /#extras -\-> -->
    
    <footer id="contentinfo" class="body">
      <address id="about" class="vcard body">
	<p>
          Generated with <a href="http://getpelican.com/">Pelican</a>, a
          static site generator which
          uses <a href="http://python.org">Python</a>.
	</p>
        <p>
	  Theme modified from the original <i>notmyidea</i> theme
	  by <a href="http://coding.smashingmagazine.com/2009/08/04/designing-a-html-5-layout-from-scratch/">Smashing
	  Magazine</a>.
	</p>
      </address><!-- /#about -->
    </footer><!-- /#contentinfo -->

    <!--  -->
    <!--  -->
    <!--  -->

  </body>
</html>