<!DOCTYPE html>
<html lang="en">
  
  <head>
    <meta charset="utf-8" />
    <title>De novo assembly - Complete workflow</title>
    <link rel="stylesheet" href="./theme/css/main.css" />
    <!--[if IE]>
        <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->
  </head>

  <body id="index" class="home">
    <header id="banner" class="body">
      <h1><a href="./">Roscoff course 2015 - RAD tags </a></h1>
      <nav><ul>
<!-- http://stackoverflow.com/questions/18520046/how-can-i-control-the-order-of-pages-from-within-a-pelican-article-category -->
          <li><a href="./index.html">About</a></li>
          <li><a href="./practicals.html">Practicals</a></li>
          <li><a href="./bibliography-notes.html">Bibliography notes</a></li>
      </ul></nav>
    </header><!-- /#banner -->

<section id="content" class="body">
    <!-- <h1 class="entry-title">De novo assembly - Complete workflow</h1> -->
    
    


<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#sec-1"><i>De novo</i> assembly</a>
<ul>
<li><a href="#sec-1-1">Assembly into consensus sequences</a>
<ul>
<li><a href="#sec-1-1-1">Concatenate all reads for a given sense</a></li>
<li><a href="#sec-1-1-2"><span class="todo TODO">TODO</span> Run CAP3 (forward reads)</a></li>
<li><a href="#sec-1-1-3"><span class="todo TODO">TODO</span> Run CAP3 (reverse reads)</a></li>
<li><a href="#sec-1-1-4">CAP3 run</a></li>
</ul>
</li>
<li><a href="#sec-1-2">Quality filtering of consensus sequences</a>
<ul>
<li><a href="#sec-1-2-1"><span class="todo TODO">TODO</span> Examine read length</a></li>
<li><a href="#sec-1-2-2"><span class="todo TODO">TODO</span> Identify repetitive motives or extremely similar consensus sequences</a></li>
<li><a href="#sec-1-2-3"><span class="todo TODO">TODO</span> Remove microsatellite repeats</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-1">
<h2 id="sec-1"><i>De novo</i> assembly</h2>
<div class="outline-text-2" id="text-1">
<p>
<div class="navLink"><a href="part-one-02-raw-reads-processing.html">Previous: Raw reads processing</a></div>
</p>
</div>
<div class="outline-3" id="outline-container-sec-1-1">
<h3 id="sec-1-1">Assembly into consensus sequences</h3>
<div class="outline-text-3" id="text-1-1">
<p>
We choose to go for a <i>de novo</i> assembly since we do not have a genome for this
species. Several tools were tested, and in the end we decided to go with a bit
older tool, CAP3 (<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC310812/">original publication</a>), which can incorporate quality
information in the assembly process.
</p>
<p>
We perform two independent assemblies, one for the forward reads and one for
the reverse reads.
</p>
</div>
<div class="outline-4" id="outline-container-sec-1-1-1">
<h4 id="sec-1-1-1">Concatenate all reads for a given sense</h4>
<div class="outline-text-4" id="text-1-1-1">
<p>
We have to pool all the reads of a given sense together:
</p>
<div class="org-src-container">
<pre class="src src-bash">ls -1
</pre>
</div>
<pre class="example">
ABB.pair.for.fastq.clean
ABB.pair.rev.fastq.clean
BYN.pair.for.fastq.clean
BYN.pair.rev.fastq.clean
HKI.pair.for.fastq.clean
HKI.pair.rev.fastq.clean
LEV.pair.for.fastq.clean
LEV.pair.rev.fastq.clean
POR.pair.for.fastq.clean
POR.pair.rev.fastq.clean
PYO.pair.for.fastq.clean
PYO.pair.rev.fastq.clean
RYT.pair.for.fastq.clean
RYT.pair.rev.fastq.clean
SKA.pair.for.fastq.clean
SKA.pair.rev.fastq.clean
</pre>
<div class="org-src-container">
<pre class="src src-bash">cat *for* &gt; for.fastq
cat *rev* &gt; rev.fastq
</pre>
</div>
<p>
We obtain two large files with all the reads.
</p>
</div>
</div>
<div class="outline-4" id="outline-container-sec-1-1-2">
<h4 id="sec-1-1-2"><span class="todo TODO">TODO</span> Run CAP3 (forward reads)</h4>
<div class="outline-text-4" id="text-1-1-2">
<p>
CAP3 takes for input two files, one containing the FASTA sequences and one
containing the corresponding quality scores. We can split the FASTQ files into
those two files using a <a href="resources/split_fasta_qual.py">python script</a> (<code>split_fasta_qual.py</code>):
</p>
<div class="org-src-container">
<pre class="src src-bash">python split_fasta_qual.py for.fastq
python split_fasta_qual.py rev.fastq
</pre>
</div>
<p>
We submit the CAP3 assembly as a job on the ABiMS server. We need to prepare
our <a href="resources/cap3-for-job">job file</a> (<code>cap3-for-job</code>) (see the ABiMS <a href="http://abims.sb-roscoff.fr/resources/cluster/howto">how-to</a> page for more details about
job files and how to submit them):
</p>
<div class="org-src-container">
<pre class="src src-bash">#!/bin/bash
#$ -S /bin/bash
#$ -M matthieu.bruneaux@ens-lyon.org
#$ -V
#$ -m bea
#$ -cwd
cap3 for.fastq.fasta &gt; for.cap3.log
</pre>
</div>
<p>
Here the command is simply <code>cap3 for.fastq.fasta &gt; for.cap3.log</code>. CAP3
automatically gets the quality information from <code>for.fastq.fasta.qual</code>.
</p>
<p>
The job file is submitted to the long queue with:
</p>
<div class="org-src-container">
<pre class="src src-bash">qsub -q long.q cap3-for-job
</pre>
</div>
</div>
</div>
<div class="outline-4" id="outline-container-sec-1-1-3">
<h4 id="sec-1-1-3"><span class="todo TODO">TODO</span> Run CAP3 (reverse reads)</h4>
<div class="outline-text-4" id="text-1-1-3">
<p>
Submit a similar assembly for the reverse reads.
</p>
<p>
An alternative approach would have been to map the reads back to the reference
genome. Why did we not use the threespine genome as a reference? 
</p>
<p>
(When mapping back we have to be quite stringent and allow only for example one
or two mismatch, and one unique good quality mapping location. With the
reference genome of a different species, this can be problematic and the
mapping back might be less efficient that creating consensus contigs or
stacks.)
</p>
</div>
</div>
<div class="outline-4" id="outline-container-sec-1-1-4">
<h4 id="sec-1-1-4">CAP3 run</h4>
<div class="outline-text-4" id="text-1-1-4">
<p>
The CAP3 run is quite long (a bit more than 15 hours for the reverse reads). We
give you the assembly for the reverse reads (<code>CAP3-rev.fasta</code>), using the
default parameters of CAP3. Note that for a real analysis we would need to made
several tests with different parameters and select the ones that give the
better assembly.
</p>
</div>
</div>
</div>
<div class="outline-3" id="outline-container-sec-1-2">
<h3 id="sec-1-2">Quality filtering of consensus sequences</h3>
<div class="outline-text-3" id="text-1-2">
<p>
The <i>de novo</i> assembly with CAP3 can be tuned by changing the parameters from
the default ones (which we should do for a real analysis). An assembly is
likely not to be perfect and the resulting consensus sequences have to be
quality checked before being used for read mapping.
</p>
</div>
<div class="outline-4" id="outline-container-sec-1-2-1">
<h4 id="sec-1-2-1"><span class="todo TODO">TODO</span> Examine read length</h4>
<div class="outline-text-4" id="text-1-2-1">
<p>
Because of the ddRAD protocol, we expect to produce consensus sequences by
piling up identical or very similar reads into stacks. Since most of the reads
have the same length to start with (xx for forward reads and xx for reverse
reads after cutting the barcode and restriction sites), we expect to obtain
consensus sequences of the same length.
</p>
<p>
We can get the length of the reads by using a homemade <a href="resources/get_read_length.py">python script</a>
(<code>get_fasta_length.py</code>):
</p>
<div class="org-src-container">
<pre class="src src-bash">python get_fasta_length.py cap3-rev.fasta
</pre>
</div>
<p>
We can plot the read length distribution with R:
</p>
<div class="org-src-container">
<pre class="src src-R"># R script
#
# Plot read length distribution
#------------------------------

d = read.table("cap3-rev.fasta.lengths")
plot(hist(d[, 1])
</pre>
</div>
<p>
How does the length distribution look like? What is the length of forward
reads? Of reverse reads? How would you remove the sequences that do not match
the median length?
</p>
</div>
</div>
<div class="outline-4" id="outline-container-sec-1-2-2">
<h4 id="sec-1-2-2"><span class="todo TODO">TODO</span> Identify repetitive motives or extremely similar consensus sequences</h4>
<div class="outline-text-4" id="text-1-2-2">
<p>
CAP3 sometimes produces consensus which contain repetitive motives or very
similar consensus sequences. Those sequences are problematic for the read
mapping back. We perform a blast of the consensus sequences against themselves
to identify sequences which have motives with many matches, or duplicate
consensus sequences.
</p>
<p>
To perform the blast search, we firt prepare a blast database from the
consensus sequences and then we use <code>blastn</code>:
</p>
<div class="org-src-container">
<pre class="src src-bash"># Prepare the blast database
makeblastdb -help
makeblastdb -in cap3-rev.fasta -dbtype nucl
</pre>
</div>
<p>
We have to perform the <code>blastn</code> by submitting a job because it is a long
process. See <code>blastn -help</code> for more details. The job file is
<code>cap3-blastn-job</code>:
</p>
<pre class="example">
#!/bin/bash
#$ -S /bin/bash
#$ -M matthieu.bruneaux@ens-lyon.org
#$ -V
#$ -m bea
#$ -cwd
# Perform the blast search (output in tabular format)
blastn -task blastn -db cap3-rev.fasta -query cap3-rev.fasta -out cap3-blastn-results -outfmt 6
</pre>
<p>
We submit the job with <code>qsub -q long.q cap3-for-job</code>. Here we ask for a tabular
output format.
</p>
<p>
Next, we parse the results to identify consensus sequences to be removed or to
be merged together. We must remember that the default output for <code>blastn</code> in
tabular format is (from <code>blastn -help</code>):
</p>
<pre class="example">
qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore
</pre>
<p>
We can first have a look at the distribution of the size of the matches;
</p>
<div class="org-src-container">
<pre class="src src-bash">cut -f 4 cap3-blastn-results &gt; lengths
</pre>
</div>
<div class="org-src-container">
<pre class="src src-R"># R script
d = read.table("lengths")[, 1]
hist(d, xlim = c(0,120), breaks = 0:120, col = "cornflowerblue")
</pre>
</div>
<p>
What is the minimal match length observed in this distribution? How would you
differentiate between repeated motives and duplicate consensus sequences?
</p>
<p>
We can also look at the number of matches for each consensus sequence:
</p>
<pre class="example">
# command line
cut -f 1 cap3-blastn-results &gt; seq-names
sort seq-names &gt; seq-names-sorted
uniq -c seq-names-sorted &gt; seq-names-count
sort -nr seq-names-count &gt; seq-match-count
</pre>
<div class="org-src-container">
<pre class="src src-R"># R script
d = read.table("seq-match-count")
names(d) = c("count", "consensus")
plot(d$count, pch = ".")
plot(d$count, pch = ".", ylim = c(0, 100))
hist(d$count, col = "cornflowerblue", xlab = "N matches")
hist(d$count, col = "cornflowerblue", ylim = c(0, 10000), xlab = "N matches")
hist(d$count, breaks = 1000, col = "cornflowerblue", xlab = "N matches",
     xlim = c(0, 200))
</pre>
</div>
<p>
How does the distribution of matches look like? How would you select a
threshold? How would you apply it?
</p>
<p>
Here is some R code to extract the names of the consensus sequences for which
the number of matches:
</p>
<pre class="example">
# R
d = read.table("seq-match-count")
names(d) = c("count", "consensus")
threshold = xx # Replace xx with your threshold for the number of matches allowed
d2 = subset(d, d$count &lt;= threshold)
nrow(d)
nrow(d2)
write.table(d2$consensus, file = "names-kept-consensus", quote = F,
  row.names = F, col.names = F)
</pre>
<p>
Have a look at <code>names-kept-consensus</code> with <code>less</code>. How would you use this file
to filter only the consensus sequences we want to keep? If you have some time,
do it.
</p>
</div>
</div>
<div class="outline-4" id="outline-container-sec-1-2-3">
<h4 id="sec-1-2-3"><span class="todo TODO">TODO</span> Remove microsatellite repeats</h4>
<div class="outline-text-4" id="text-1-2-3">
<p>
The last quality control step is to remove consensus sequences that contains
microsatellite repeats. To do this, we use Sputnik on the consensus
sequences. We should use it on the consensus sequence file after removing the
sequences with repeated motives, but since the intermediate file is missing
we'll run it on the full consensus file. We have to submit a job again,
<code>sputnik-job</code>:
</p>
<pre class="example">
#!/bin/bash
#$ -S /bin/bash
#$ -M matthieu.bruneaux@ens-lyon.org
#$ -V
#$ -m bea
#$ -cwd
# Perform the blast search (output in tabular format)
sputnik cap3-rev.fasta
</pre>
<p>
We submit it with <code>qsub -q long.q sputnik-job</code>.
</p>
<p>
Examine the output of Sputnik and count the number of contigs with
microsatellite repeats. How would you proceed to remove them from the initial
consensus file?
</p>
<p>
We will now give you two fasta files containing the filtered consensus
sequences that we decide to keep in the end.
</p>
<p>
<div class="navLink"><a href="part-one-04-variant-calling-genotyping.html">Next: Variant calling and genotyping</a></div>
</p>
</div>
</div>
</div>
</div>

</section>

    <!-- <section id="extras" class="body"> -->
    <!--  -->
    <!--  -->
    <!-- </section><\!-- /#extras -\-> -->
    
    <footer id="contentinfo" class="body">
      <address id="about" class="vcard body">
	<p>
          Generated with <a href="http://orgmode.org/">Org mode</a>
          in <a href="http://www.gnu.org/software/emacs/">Emacs</a>
          and <a href="http://getpelican.com/">Pelican</a>, a static site
          generator which uses <a href="http://python.org">Python</a>.
	</p>
        <p>
	  Theme modified from the original <i>notmyidea</i> theme
	  by <a href="http://coding.smashingmagazine.com/2009/08/04/designing-a-html-5-layout-from-scratch/">Smashing
	  Magazine</a>.
	</p>
      </address><!-- /#about -->
    </footer><!-- /#contentinfo -->

    <!--  -->
    <!--  -->
    <!--  -->

  </body>
</html>