<html>
<head>
<meta name="Title" content="Practicals" />
<meta name="Summary" content="Practicals" />
<meta name="URL" content="practicals.html" />
<meta name="Save_as" content="practicals.html" />
<meta name="Sortorder" content="015" />
<meta name="Slug" content="practicals" />
</head>
<body>


<div class="outline-2" id="outline-container-sec-1">
<h2 id="sec-1">Plan</h2>
<div class="outline-text-2" id="text-1">
<p>
The practicals are done in two parts:
</p>
<ul class="org-ul">
<li>A <b>detailed analysis</b>, from raw reads to final results, of a dataset from a
paper published in 2013: <a href="http://onlinelibrary.wiley.com/doi/10.1111/j.1365-294X.2012.05749.x/abstract">Molecular evolutionary and population genomic analysis of the 
nine-spined stickleback using a modified restriction-site-associated DNA tag approach</a>. 
The raw dataset is available from the Dryad data archive, doi:
<a href="http://dx.doi.org/10.5061/dryad.44sf1">10.5061/dryad.44sf1</a>. We will try to <b>reproduce</b> the results from the paper,
<b>criticize</b> the different steps and maybe <b>try different approaches</b> as we
go.

<p>
After this part, you should have a general view of a RAD-seq workflow.
</p>
</li>
<li>A <b>more specific look</b> at some precise steps of the analysis, using other
published datasets. This might entail, for example, genotype calling using a
maximum likelihood approach.
</li>
</ul>
</div>
</div>
<div class="outline-2" id="outline-container-sec-2">
<h2 id="sec-2">Detailed instructions</h2>
<div class="outline-text-2" id="text-2">
<p>
During the practicals, we will be using an Etherpad document to share code. You
can access the Etherpad document <a href="https://etherpad.fr/p/1DAjASHRdI">here</a>.
</p>
<p>
Please follow the links to proceed to the detailed instructions:
</p>
<ul class="org-ul">
<li><b><a href="part-one-01-introduction.html">Part one: a complete analysis, from raw reads to results</a></b>
<ul class="org-ul">
<li><a href="part-one-01-introduction.html">Introduction</a>
</li>
<li><a href="part-one-02-raw-reads-processing.html">Raw reads processing</a>
</li>
<li><a href="part-one-03-de-novo-assembly.html"><i>De novo</i> assembly</a>
</li>
<li><a href="part-one-04-variant-calling-genotyping.html">Variant calling and genotyping</a>
</li>
<li><a href="part-one-05-downstream-analysis.html">Downstream analysis</a>
<ul class="org-ul">
<li><a href="part-one-06-phylogeographic-tree.html">Phylogeographic tree</a>
</li>
<li><a href="part-one-07-mapping-back-3sp-genome.html">Mapping back to the threespine genome</a>
</li>
<li><a href="part-one-08-prediction-SNP-effect.html">Prediction of SNP effects in coding regions</a>
</li>
<li><a href="part-one-09-genome-scan.html">Genome scan for G<sub>st</sub></a>
</li>
<li><a href="part-one-10-GO-enrichment-tests.html">Gene ontology (GO) enrichment tests</a>
</li>
</ul>
</li>
<li><a href="part-one-11-conclusion.html">Conclusion</a>
</li>
</ul>
</li>
<li><b><a href="part-two-01-introduction.html">Part two: going further for some analytical steps</a></b>
</li>
</ul>
</div>
</div>
</body></html>
